cirq/protocols/apply_unitary.py

```diff
@@ -227,27 +264,48 @@ def apply_unitary(unitary_value: Any,
     if func is not None:
         result = func(args)
         if result is not NotImplemented and result is not None:
+            if result is args.target_tensor:
+                return result
+            # If any entries of unitary_value's shape are less than the
+            # corresponding entries of args.qid_shape, don't trust func to
+            # copy the untouched entries from args.target_tensor into result.
+            op_shape = qid_shape_protocol.qid_shape(unitary_value,
+                                                    (2,) * len(args.axes))
+            if any(op_shape[i] < args.qid_shape[axis]
+                   for i, axis in enumerate(args.axes)):
+                # Copy extra entries from args.target_tensor into result.
+                for i in range(len(args.axes)):
+                    axis = args.axes[i]
+                    op_d = op_shape[i]
+                    axis_d = args.qid_shape[axis]
```

Comment:
Make sure to only copy each place in the array once (e.g. don't copy (2:, :) and also (:, 2:) because this does (2:,2:) twice).

Reply:
Fixed with new args slicing method.